### ISSUE TITLE
TypeScript usage with render prop components?

### ISSUE THREAD

**Issue by FredyC** [apollographql/react-apollo#1874]
This is more like a question or a discussion. The core idea is that I am using [the code generator](https://github.com/dotansimha/graphql-code-generator) to get typescript definitions for my queries/mutations. To use those I had built custom render prop component a while ago wrapping around the `graphql` HoC. In the end, it looks like this...

```ts
import gql from 'graphql-tag'

import { GLogoutWidget } from 'graph/types'

export const LogoutWidgetQuery = buildQuery<
  GLogoutWidget.Query,
  GLogoutWidget.Variables
>(gql`
  query GLogoutWidget($id: ID!) {
    user(id: $id) {
      id
      firstName
      lastName
    }
  }
`)
```

This works really well and can be used just as easily as a new `Query` component except it is fully typed.

```tsx
export const LogoutWidget: React.SFC = ({ userId }) => (
  <LogoutWidgetQuery
    variables={{ id: userId }}
    render={({ user }) => (
      <Link to="/logout">{`${user.firstName} ${user.lastName}`}</Link>
    )}
  />
)
```

---

Now I am wondering if I can achieve something similiar with a new `Query` component? I can only think about a following approach, but I don't like that as I want to keep types separated from a component use.

```tsx
export const LogoutWidget: React.SFC = ({ userId }) => (
  <Query
    variables={{ id: userId }: GLogoutWidget.Variables}
    render={({ user }: GLogoutWidget.Query) => (
      <Link to="/logout">{`${user.firstName} ${user.lastName}`}</Link>
    )}
  />
)
```

I was thinking about the approach like this, but that's giving cryptic error: `Type 'typeof Query' is not assignable to type 'Query<Query, Variables>'.
  Property 'context' is missing in type 'typeof Query'.` I am not sure if it can be tricked somehow into this.

```ts
const LogoutWidgetQuery: Query<
  GLogoutWidget.Query,
  GLogoutWidget.Variables
> = Query
```

**Comment by jagregory** [comment]
This should do the trick:

```js
class LogoutWidgetQuery extends Query<GLogoutWidget.Query, GLogoutWidget.Variables> {}
```

**Comment by FredyC** [comment]
@jagregory Thanks, that's surely a step in a right direction, but not exactly the same considering that my first example had typings *and* query document included. Unfortunately, that does not work simply because `props` here are typed `any` and all type sweetness is lost down the lane.

```tsx
class QueryClass extends Query<GLogoutWidget.Query, GLogoutWidget.Variables> {}

export const LogoutWidgetQuery = (props) => <QueryClass query={gql`...`} {...props} />
```

Instead, I came up with this solution that's pretty similar to the one I already have. I am more fan of a `render` prop than _abusing children_. It seems to be working just fine.

```tsx
function buildQuery<TData, TVariables>(query: DocumentNode) {
  interface IProps extends QueryProps<TData, TVariables> {
    render(data: QueryResult<TData, TVariables>): ReactNode
  }
  return class extends React.Component<IProps> {
    render() {
      const { render, ...props } = this.props
      return (
        <Query query={query} {...props}>
          {render}
        </Query>
      )
    }
  }
```

It could be even more simpler without classes, unfortunately there is a long standing issue [that is being worked](https://github.com/Microsoft/TypeScript/issues/21699) on but won't be fixed till TypeScript 2.9. 

```tsx
function buildQuery<TData, TVariables>(query: DocumentNode) {
  interface IProps extends QueryProps<TData, TVariables> {
    render(data: QueryResult<TData, TVariables>): ReactNode
  }
  return ({ render, ...props }): React.SFC<IProps> (
   <Query query={query} {...props}>
     {render}
   </Query>
  )
}
```

Anyway, I am satisfied with this for now. I was kinda afraid that with this move toward render prop components, the `graphql` HoC will become deprecated and I would lost ability to use this approach.

**Comment by deltaskelta** [comment]
@FredyC even when I use @jagregory 's solution without passing props in as well. I still seem to lose all types down the line. The render prop comes through as a `QueryResult<any>`.

I am looking for some patterns that can let me use the generated types as well, but they seem rather pointless if the render prop data just comes through as `any`
